homeassistant/components/pglab/config_flow.py

```diff
@@ -0,0 +1,61 @@
+"""Config flow for PG LAB Electronics integration."""
+
+from __future__ import annotations
+
+from typing import Any
+
+from homeassistant.config_entries import ConfigFlow, ConfigFlowResult
+from homeassistant.helpers.service_info.mqtt import MqttServiceInfo
+
+from .const import DISCOVERY_TOPIC, DOMAIN
+
+CONF_DISCOVERY_PREFIX = "discovery_prefix"
+
+
+class FlowHandler(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow."""
+
+    VERSION = 1
+
+    async def async_step_mqtt(
+        self, discovery_info: MqttServiceInfo
+    ) -> ConfigFlowResult:
+        """Handle a flow initialized by MQTT discovery."""
+        if self._async_in_progress() or self._async_current_entries():
+            return self.async_abort(reason="single_instance_allowed")
+
+        await self.async_set_unique_id(DOMAIN)
+
+        # Validate the message, abort if it fails.
+        if not discovery_info.topic.endswith("/config"):
+            # Not a PGLab Electronics discovery message.
+            return self.async_abort(reason="invalid_discovery_info")
+        if not discovery_info.payload:
+            # Empty payload, Unexpected payload.
+            return self.async_abort(reason="invalid_discovery_info")
+
+        return await self.async_step_confirm()
+
+    async def async_step_user(
```

Comment:
Is there a way we can do a connection check in the user step? How can we confirm everything works? the best way to tell the user it won't work is via the config flow

Reply:
Not really sure. The PGLab device (E-Board) communicate with MQTT server only.  When E-Board start, it publish a  MQTT HA discovery message ...  if HA is able to find the discovery message I suppose that "every thing works properly".
After the integration setup, the user can turn On-Off switches and check if E-Board is able to receive MQTT message from HA.  To have a more strong check probably I  have to introduce a sort of hand-shaking mechanism between HA and E-Board firmware... For this patch I don't want to do it because I have to review E-Board firmware as well. Eventually I'll do in a different patch.

Reply:
I discussed this with some others and we think that the user step isn't logical if the device still works via discovery. The step would only allow new users to think they have set it up correctly, even though there might be other problems. I would suggest looking at what `drop_connect` did with their approach on the user step.

Reply:
I'll do as you suggested... but I don't see a big problem to add PGLab integration for user configuration.
If no E-Board device is connected to MQTT server, the PGLab integration don't expose nothing, it stay in the "background" wait a discovery message. When E-Board add a MQTT discovery message, the PGLab integration automatically add the new device with the supported switch entities. In some way you can setup HA to use PGLab devices before to physically connect E-Board to the network.

Reply:
In the old state, when you debug an user, you still need more information if the user did it manually or via discovery to know if it ever worked in the first place. In the new state, you are guaranteed that it did work at some point, which helps your debug speed and increases the user experience.

Reply:
I don't really agree with @joostlek here, I think it's confusing to not have a manual flow. The problem with discovery flows is that they can be ignored and the user then no longer sees them, and there's no obvious way for the user to add the integration.

Reply:
This is good point.  But I'm not really sure how I should change the code.
Restore as before ?

Reply:
I think this is fine in the current state; the user step was added back by 6ab23210189e3b4b017f9257d1b61125d82b6e19